src/TodoByTicketRule.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+namespace staabm\PHPStanTodoBy;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Node\VirtualNode;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use staabm\PHPStanTodoBy\utils\TicketStatusFetcher;
+use function preg_match_all;
+use function substr_count;
+use function trim;
+use const PREG_OFFSET_CAPTURE;
+use const PREG_SET_ORDER;
+
+/**
+ * @implements Rule<Node>
+ */
+final class TodoByTicketRule implements Rule
+{
+    private const PATTERN = <<<'REGEXP'
+/
+@?TODO # possible @ prefix
+@?[a-zA-Z0-9_-]*\s* # optional username
+\s*[:-]?\s* # optional colon or hyphen
+(?P<ticket>[A-Z]+-\d+) # ticket id consisting of ABC-123 format
+\s*[:-]?\s* # optional colon or hyphen
+(?P<comment>.*) # rest of line as comment text
+/ix
+REGEXP;
+
+    private bool $nonIgnorable;
+    /** @var list<non-empty-string> */
+    private array $resolvedStatuses;
+    private string $keyPrefix;
+    private TicketStatusFetcher $fetcher;
+
+    /** @param list<non-empty-string> $resolvedStatuses */
+    public function __construct(bool $nonIgnorable, array $resolvedStatuses, string $keyPrefix, TicketStatusFetcher $fetcher)
+    {
+        $this->nonIgnorable = $nonIgnorable;
+        $this->resolvedStatuses = $resolvedStatuses;
+        $this->keyPrefix = $keyPrefix;
+        $this->fetcher = $fetcher;
+    }
+
+    public function getNodeType(): string
+    {
+        return Node::class;
+    }
+
+    public function processNode(Node $node, Scope $scope): array
```

Comment:
the `processNode` functions of all rules are pretty redundant atm. let me land a preparing refactoring on the main branch so we can remove duplication

Reply:
refactored with https://github.com/staabm/phpstan-todo-by/pull/28 and https://github.com/staabm/phpstan-todo-by/pull/27 - please rebase